### PR TITLE
[Youtube] Support additional formats from YouTube Music when requested

### PR DESCRIPTION
**Pull request by colethedj**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [ ] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [X] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [X] Improvement
- [ ] New extractor
- [X] New feature

---

### Description of your *pull request* and other information

Explanation of your *pull request* in arbitrary form goes here. Please make sure the description explains the purpose and effect of your *pull request* and is worded well enough to be understood. Provide as much context and examples as possible.

From https://github.com/ytdl-org/youtube-dl/pull/28778 which in turn is from https://github.com/ytdl-org/youtube-dl/pull/26160
TL;DR, adds support for format 141 and any other YTM-only formats that come about.

@craftingmod let us know if you are happy with us merging this here.

Changes from original PR:
- Use URL smuggling to handle YTM playlists more elegantly without breaking `--playlist-end` etc (credit @pukkandan).
- Only use YTM data for additional formats (i.e. merge the formats with normal Youtube player res formats). This should fix any extraction issues due to differences between the YTM and YT player responses.
- Added a test

TODO:
- investigate 404s (possibly upcoming API deprecation)

### CODE REVIEW

File: yt_dlp/extractor/youtube.py [hunk 1]
```diff
@@ -1847,11 +1851,47 @@ def _extract_comments(self, ytcfg, video_id, contents, webpage, xsrf_token):
...
         video_id = self._match_id(url)
+
+        is_music_url = smuggled_data.get('is_music_url') or self.is_music_url(url)
+
         base_url = self.http_scheme() + '//www.youtube.com/'
         webpage_url = base_url + 'watch?v=' + video_id
         webpage = self._download_webpage(
             webpage_url + '&bpctr=9999999999&has_verified=1', video_id, fatal=False)
 
+        # define early for use
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:1862]:
```suggestion
```

**Comment by craftingmod** [comment]
Thanks to amazing works! 👍 

My python skill is so bad not enough to implement all 😭

**Comment by Chartman123** [comment]
@pukkandan Am I doing something wrong? In contrast to @craftingmod's original code I can only download the lower quality files with your branch... I'm using the same cookies for login with an YouTube Music Premium account.

**Comment by pukkandan** [comment]
@Chartman123 Can you show the output of `yt-dlp -v -F --cookies cookiefile URL`

**Comment by Chartman123** [comment]
My URL is a playlist URL, so the -F flag won't show any outpout regarding the available formats.

**Comment by pukkandan** [comment]
It does...
Anyway, use this: `yt-dlp -v -F --cookies cookiefile --playlist-end 1 URL` since I only need to see the formats of 1 video

**Comment by Chartman123** [comment]
````
[debug] User config file: /home/chris/.config/yt-dlp/config
[debug] User config: ['-x', '-o', '~/Downloads/%(playlist)s/%(title)s_%(id)s.%(ext)s', '--cookies', '~/.config/yt-dlp/cookies.txt', '--download-archive', '~/.config/yt-dlp/archive.txt', '--no-mtime', '--audio-quality', '0', '--add-metadata', '--embed-thumbnail']
[debug] Command-line config: ['https://music.youtube.com/playlist?list=PLk73otWzxw9mb2Ob1lAOfg5cprIy8yHcP', '-v', '-F', '--playlist-end', '1']
[debug] Loading archive file '/home/chris/.config/yt-dlp/archive.txt'

[debug] Encodings: locale UTF-8, fs utf-8, out utf-8, pref UTF-8
[debug] yt-dlp version 2021.05.11 (source)
[debug] Plugin Extractors: ['SamplePlugin']
[debug] Python version 3.8.9 (CPython 64bit) - Linux-5.12.0-2-default-x86_64-with-glibc2.2.5
[debug] exe versions: ffmpeg 4.4, ffprobe 4.4
[debug] Proxy map: {}
[youtube:tab] PLk73otWzxw9mb2Ob1lAOfg5cprIy8yHcP: Downloading webpage
[youtube:tab] PLk73otWzxw9mb2Ob1lAOfg5cprIy8yHcP: Downloading API JSON with unavailable videos
[download] Downloading playlist: Download
[youtube:tab] playlist Download: Downloading 1 videos
[download] Downloading video 1 of 1
[youtube] sotQsDzrwkE: Downloading webpage
[youtube] sotQsDzrwkE: Fetching youtube-music info webpage
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, vcodec:vp9.2(10), acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[info] Available formats for sotQsDzrwkE:
ID  EXT  RESOLUTION FPS |  FILESIZE    TBR PROTO | VCODEC        VBR ACODEC     ABR  ASR    NOTE
--- ---- ---------- --- - ---------- ----- ----- - ----------- ----- --------- ---- ------- ------------------------
140 m4a  audio only     |  2.89MiB    129k https |                   mp4a.40.2 129k 44100Hz tiny, m4a_dash, 44100Hz
141 m4a  audio only     |  5.74MiB    257k https |                   mp4a.40.2 257k 44100Hz tiny, m4a_dash, 44100Hz
249 webm audio only     |  1.26MiB     56k https |                   opus       56k 48000Hz tiny, webm_dash, 48000Hz
250 webm audio only     |  1.64MiB     73k https |                   opus       73k 48000Hz tiny, webm_dash, 48000Hz
251 webm audio only     |  3.17MiB    142k https |                   opus      142k 48000Hz tiny, webm_dash, 48000Hz
160 mp4  144x144    25  |  297.28KiB   13k https | avc1.4d400b   13k                        144p, mp4_dash
278 webm 144x144    25  |  329.52KiB   14k https | vp9           14k                        144p, webm_dash
133 mp4  240x240    25  |  536.28KiB   23k https | avc1.4d400c   23k                        240p, mp4_dash
242 webm 240x240    25  |  380.72KiB   16k https | vp9           16k                        240p, webm_dash
134 mp4  360x360    25  |  948.60KiB   41k https | avc1.4d4015   41k                        360p, mp4_dash
18  mp4  360x360    25  |  3.06MiB    137k https | avc1.42001E  137k mp4a.40.2   0k 44100Hz 360p, 44100Hz
243 webm 360x360    25  |  539.75KiB   23k https | vp9           23k                        360p, webm_dash
135 mp4  480x480    25  |  1.80MiB     80k https | avc1.4d401e   80k                        480p, mp4_dash
244 webm 480x480    25  |  779.80KiB   34k https | vp9           34k                        480p, webm_dash
136 mp4  720x720    25  |  5.13MiB    230k https | avc1.64001f  230k                        720p, mp4_dash
247 webm 720x720    25  |  2.05MiB     91k https | vp9           91k                        720p, webm_dash
137 mp4  1080x1080  25  |  13.94MiB   626k https | avc1.640020  626k                        1080p, mp4_dash
248 webm 1080x1080  25  |  10.16MiB   456k https | vp9          456k                        1080p, webm_dash
[download] Finished downloading playlist: Download
````
Format 141 is in the list but doesn't get downloaded

**Comment by fipsingdale** [comment]
i keep getting: 
`[WARNING] ... at YoutubeDL.py:757 -- [youtube] unable to download youtube music info webpage: HTTP Error 404: Not Found
`
... does anyone have an idea what might be causing this?

**Comment by pukkandan** [comment]
@fipsingdale Open a new issue with all relevent info including verbose log

**Comment by colethedj** [comment]
> i keep getting: 
> `[WARNING] ... at YoutubeDL.py:757 -- [youtube] unable to download youtube music info webpage: HTTP Error 404: Not Found
> `
> ... does anyone have an idea what might be causing this?

update to latest release.
